ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,573 @@
+import org.apache.helix.HelixManager;
+import org.apache.helix.HelixManagerFactory;
+import org.apache.helix.InstanceType;
+import org.apache.helix.LiveInstanceChangeListener;
+import org.apache.helix.NotificationContext;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.LiveInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.ClusterMapUtils.*;
+
+
+/**
+ * An implementation of {@link ClusterMap} that makes use of Helix to dynamically manage the cluster information.
+ *
+ * @see <a href="http://helix.apache.org">http://helix.apache.org</a>
+ */
+class HelixClusterManager implements ClusterMap {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final String clusterName;
+  private final MetricRegistry metricRegistry;
+  private final ClusterMapConfig clusterMapConfig;
+  private final Map<String, DcZkInfo> dcToDcZkInfo = new HashMap<>();
+  private final Map<String, AmbryPartition> partitionNameToPartitionIdMap = new ConcurrentHashMap<>();
+  private final Map<String, AmbryDataNode> instanceNameToDataNodeIdMap = new ConcurrentHashMap<>();
+  private final Map<AmbryPartition, Set<AmbryReplica>> partitionIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryReplica>> dataNodeIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryDisk>> dataNodeIdToDiskIds = new ConcurrentHashMap<>();
+  private final Map<ByteBuffer, AmbryPartition> partitionMap = new ConcurrentHashMap<>();
+  private final HelixClusterManagerCallback helixClusterManagerCallback;
+  private long clusterWideRawCapacityBytes;
+  private long clusterWideAllocatedRawCapacityBytes;
+  private long clusterWiseAllocatedUsableCapacityBytes;
+  final HelixClusterManagerMetrics helixClusterManagerMetrics;
+
+  /**
+   * Instantiate a HelixClusterManager.
+   * @param clusterMapConfig the {@link ClusterMapConfig} associated with this manager.
+   * @param instanceName the String representation of the instance associated with this manager.
+   * @throws IOException if there is an error in parsing the clusterMapConfig or in connecting with the associated
+   *                     remote Zookeeper services.
+   */
+  HelixClusterManager(ClusterMapConfig clusterMapConfig, String instanceName) throws IOException {
+    this.clusterMapConfig = clusterMapConfig;
+    clusterName = clusterMapConfig.clusterMapClusterName;
+    metricRegistry = new MetricRegistry();
+    helixClusterManagerCallback = new HelixClusterManagerCallback();
+    helixClusterManagerMetrics = new HelixClusterManagerMetrics(metricRegistry, helixClusterManagerCallback);
+    try {
+      Map<String, String> dataCenterToZkAddress =
+          parseZkJsonAndPopulateZkInfo(clusterMapConfig.clusterMapDcsZkConnectStrings);
+      for (Map.Entry<String, String> entry : dataCenterToZkAddress.entrySet()) {
```

Comment:
Yes. Helix does not yet have the functionality of relaying cross-colo notifications.